File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/VB/MasterDetail/MasterDetailControl.vb
Public Class MasterDetailControl
  Inherits BaseControl

  Public Sub New()
    MyBase.New("MasterDetail")
    InitializeComponent()
  End Sub

  Friend Overrides Sub OpenClick()
    Dim companies = From c In CrmDataContext.Companies _
                      Select c

      bsCustomers.DataSource = companies
    MyBase.OpenClick()
  End Sub

  Friend Overrides Sub ControlsEnabled(ByVal state As Boolean)

    Me.tsbUpdate.Enabled = state
    If (state = True) Then
      If (dgCustomer.Columns.Count <> 0 Or dgOrders.Columns.Count <> 0) Then
          Return
      End If
      Me.dgCustomer.Columns.AddRange(New System.Windows.Forms.DataGridViewColumn() {Me.CompanyIDDataGridViewTextBoxColumn, Me.CompanyNameDataGridViewTextBoxColumn, Me.PrimarycontactDataGridViewTextBoxColumn, Me.WebDataGridViewTextBoxColumn, Me.EmailDataGridViewTextBoxColumn, Me.AddresstitleDataGridViewTextBoxColumn, Me.AddressDataGridViewTextBoxColumn, Me.CityDataGridViewTextBoxColumn, Me.RegionDataGridViewTextBoxColumn, Me.PostalcodeDataGridViewTextBoxColumn, Me.CountryDataGridViewTextBoxColumn, Me.PhoneDataGridViewTextBoxColumn, Me.FaxDataGridViewTextBoxColumn})
      Me.dgOrders.Columns.AddRange(New System.Windows.Forms.DataGridViewColumn() {Me.orderIDDataGridViewTextBoxColumn, Me.companyIDDataGridViewTextBoxColumn1, Me.contactIDDataGridViewTextBoxColumn, Me.orderDateDataGridViewTextBoxColumn, Me.freightDataGridViewTextBoxColumn, Me.shipDateDataGridViewTextBoxColumn, Me.discountDataGridViewTextBoxColumn})
    Else
      dgCustomer.CancelEdit()
      dgOrders.CancelEdit()
      dgCustomer.Columns.Clear()
      dgOrders.Columns.Clear()
    End If
  End Sub

  Private Sub tsbUpdate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles tsbUpdate.Click
    Try
      If Not (CrmDataContext Is Nothing) Then
        CrmDataContext.SubmitChanges()
      End If
    Catch ex As Exception
      MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    End Try
  End Sub

  Private Sub bsOrders_ListChanged(ByVal sender As System.Object, ByVal e As System.ComponentModel.ListChangedEventArgs) Handles bsOrders.ListChanged
    If (e.ListChangedType = System.ComponentModel.ListChangedType.ItemChanged AndAlso e.PropertyDescriptor.Name = "Company") Then
      Dim curentOrder As CrmDataContext.Order = TryCast(bsOrders.List(e.NewIndex), CrmDataContext.Order)
      If (curentOrder.Company Is Nothing) Then
        CrmDataContext.Orders.DeleteOnSubmit(curentOrder)
      End If
    End If
  End Sub

  Private Sub dgCustomer_DataError(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DataGridViewDataErrorEventArgs) Handles dgCustomer.DataError, dgOrders.DataError

    If TryCast(e.Exception, ArgumentException) IsNot Nothing Then
      MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    Else
      MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error)
    End If
  End Sub
End Class

File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/CS/MasterDetail/MasterDetailControl.cs
using System.Linq;
using System;
using System.Windows.Forms;

namespace LinqDemo.MasterDetail {

  public partial class MasterDetailControl : BaseControl {

    public MasterDetailControl()
      : base("MasterDetail") {
      InitializeComponent();
    }

    internal override event EventHandler UpdateChanged;

    internal override void OpenClick() {

      var companies = from c in crmDataContext.Companies
                      select c;

      bsCustomers.DataSource = companies;
      base.OpenClick();
    }

    internal override void CloseClick() {

      bsCustomers.DataSource = typeof(CrmDataContext.Company);
      base.CloseClick();
    }

    internal override void ControlsEnabled(bool state) {

      tsbUpdate.Enabled = state;
      if (state) {
        if (dgCustomer.Columns.Count != 0 || dgOrders.Columns.Count != 0)
          return;
        dgCustomer.Columns.AddRange(new System.Windows.Forms.DataGridViewColumn[] {
            companyIDDataGridViewTextBoxColumn,
            companyNameDataGridViewTextBoxColumn,
            webDataGridViewTextBoxColumn,
            emailDataGridViewTextBoxColumn,
            addressDataGridViewTextBoxColumn,
            cityDataGridViewTextBoxColumn,
            regionDataGridViewTextBoxColumn,
            countryDataGridViewTextBoxColumn,
            phoneDataGridViewTextBoxColumn,
            faxDataGridViewTextBoxColumn});

        dgOrders.Columns.AddRange(new System.Windows.Forms.DataGridViewColumn[] {
            orderIDDataGridViewTextBoxColumn,
            companyIDDataGridViewTextBoxColumn1,
            contactIDDataGridViewTextBoxColumn,
            orderDateDataGridViewTextBoxColumn,
            freightDataGridViewTextBoxColumn,
            shipDateDataGridViewTextBoxColumn,
            discountDataGridViewTextBoxColumn});
      }
      else {
        dgCustomer.CancelEdit();
        dgOrders.CancelEdit();
        dgCustomer.Columns.Clear();
        dgOrders.Columns.Clear();
      }
    }

    private void tsbUpdate_Click(object sender, System.EventArgs e) {

      try {
        if (crmDataContext != null) {
          crmDataContext.SubmitChanges();
          if (UpdateChanged != null)
            UpdateChanged("Updated", null);
          }
      }
      catch (Exception ex) {
        MessageBox.Show(ex.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
        UpdateChanged("Update failed", null);
      }
    }

    private void bsOrders_ListChanged(object sender, System.ComponentModel.ListChangedEventArgs e) {

      if (e.ListChangedType == System.ComponentModel.ListChangedType.ItemChanged && e.PropertyDescriptor.Name == "Company") {
        CrmDataContext.Order curentOrder = (CrmDataContext.Order)bsOrders.List[e.NewIndex];
        if (curentOrder.Company == null)
          crmDataContext.Orders.DeleteOnSubmit(curentOrder);
      }
    }

    private void dgOrders_DataError(object sender, System.Windows.Forms.DataGridViewDataErrorEventArgs e) {

      if (e.Exception is ArgumentException)
        MessageBox.Show("The value is not valid.", MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
      else
        MessageBox.Show(e.Exception.Message, MainForm.SampleName, MessageBoxButtons.OK, MessageBoxIcon.Error);
    }
  }
}
